### PR TITLE
Technical feasibility k8 Secret for AuthToken

### PR DESCRIPTION
**Pull request by nmvk**
- Create a new SecretRef type
- Allow field config to declare a given field as secret
- Methods to allow fetching of secrets from k8 secrets.
- Automatic code generation to get the secret in sdk.go

Does not implement watcher which would be needed for the
update of an API.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: templates/apis/types.go.tpl [hunk 1]
```diff
@@ -14,3 +14,15 @@ import (
...
+// SecretReference has enough information to retrieve secret
+// in any namespace.
+type SecretReference struct {
+    // Namespace defines the space within which the secret name must be unique.
+    Namespace string `json:"namespace,omitempty"`
+    // Name of secret in a given namespace.
+    Name string `json:"name,omitempty"`
+    // Key of the secret.
+    Key string `json:"key,omitempty"`
+}
```
Comment (jaypipes) [inline templates/apis/types.go.tpl:27]: this should be in the ACK core types (i.e. github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1). See my comment from above.

There should not be a need for a .HasSecret boolean field of the CRD.

File: pkg/runtime/reconciler.go [hunk 2]
```diff
@@ -84,12 +86,34 @@ func (r *reconciler) BindControllerManager(mgr ctrlrt.Manager) error {
 }
 
 // SecretValueFromReference fetches the value of a Secret given a
-// SecretReference
+// namespace, name and key of the secret
 func (r *reconciler) SecretValueFromReference(
-	ref *corev1.SecretReference,
```
Comment (jaypipes) [inline pkg/runtime/reconciler.go:89]: We should have an ACK core v1alpha1 type called `SecretKeyReference` that wraps the k8s `corev1.SecretReference` with an additional `Key` field.

Place this in a `github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1/secret.go` file:

```go
import (
    k8scorev1 "k8s.io/api/core/v1"
)

// SecretKeyReference combines a k8s corev1.SecretReference with a
// specific key within the referred-to Secret
type SecretKeyReference struct {
    k8scorev1.SecretReference
    // Key is the key within the secret
    Key string `json:"key"`
}
```

**Review by jaypipes — CHANGES_REQUESTED** [review]
Hi @nmvk and @kumargauravsharma, please see my comments inline. This needs to be made to the github.com/aws-controllers-k8s/runtime repository to start and then we'll get to the changes that need to be made to the code-generator repo.

**Comment by nmvk** [comment]
Closing this, below PRs have updated information
 
- https://github.com/aws-controllers-k8s/runtime/pull/5
- https://github.com/aws-controllers-k8s/code-generator/pull/21
- https://github.com/aws-controllers-k8s/elasticache-controller/pull/9

**Pull request closed without merging**